### PR TITLE
Hyper-v: Added 2 Storage TCs

### PR DESCRIPTION
**Pull request by asofron**
Storage	STOR-22	Take and Revert Back Snapshot of VM on VHDX
Storage	STOR-23	Export/Import VM

 VHD under test **Centos_7.5_x64.vhd**
 ARM Image under test  :  :  : 
 Total Executed TestCases        2 (       2 Pass,        0 Fail,        0 Abort)
 Total Execution Time(dd:hh:mm) 0:0:12
 XML file: TestConfiguration
 
 #Sr. Test Name : Test Result : Test Duration (in minutes)
 ----------------------------------------------------
 1. STORAGE-TAKE-REVERT-SNAPSHOT : PASS : 4.49
 2. STORAGE-EXPORT-IMPORT-VM : PASS : 8.09

 VHD under test **ubuntu_18.04.vhd**
 ARM Image under test  :  :  : 
 Total Executed TestCases        2 (       2 Pass,        0 Fail,        0 Abort)
 Total Execution Time(dd:hh:mm) 0:0:13
 XML file: TestConfiguration
 
 #Sr. Test Name : Test Result : Test Duration (in minutes)
 ----------------------------------------------------
 1. STORAGE-TAKE-REVERT-SNAPSHOT : PASS : 5.54
 2. STORAGE-EXPORT-IMPORT-VM : PASS : 8.19

### CODE REVIEW

File: Testscripts/Linux/STOR-Lis-Disk.sh [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+. utils.sh || {
+    echo "Error: unable to source utils.sh!"
+    exit 0
+}
+#
+# Source constants file and initialize most common variables
+#
+UtilsInit
+
+Integrity-Check(){
```
Comment (chvalean) [inline Testscripts/Linux/STOR-Lis-Disk.sh]: add space after ) and {
do indentation of code inside function

File: Testscripts/Linux/STOR-Lis-Disk.sh [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+for driveName in /dev/sd*[^0-9];
+do
+    #
+    # Skip /dev/sda and /dev/sdb
+    #
+    if [ ${driveName} = "/dev/sda" ]; then
+        continue
+    fi
+
+    fdisk -l $driveName > fdisk.dat 2> /dev/null
```
Comment (chvalean) [inline Testscripts/Linux/STOR-Lis-Disk.sh]: would this not use the test disk that Shital is adding to VMs?

File: Testscripts/Linux/STOR-Lis-Disk.sh [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    # Format the Disk and Create a file system , Mount and create file on it .
+    (echo d;echo;echo w)|fdisk  $driveName
+    (echo n;echo p;echo 1;echo;echo;echo w)|fdisk  $driveName
+    if [ "$?" = "0" ]; then
+    sleep 5
+
+   # IntegrityCheck $driveName
+    mkfs.ext3  ${driveName}1
```
Comment (chvalean) [inline Testscripts/Linux/STOR-Lis-Disk.sh]: use ext4

File: Testscripts/Linux/STOR-Lis-Disk.sh [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            exit 0
+        fi
+    else
+        LogMsg "Error in executing fdisk  ${driveName}1"
+         SetTestStateFailed
+        exit 0
+    fi
+
+    # Perform Data integrity test
+
```
Comment (chvalean) [inline Testscripts/Linux/STOR-Lis-Disk.sh]: remove blank line

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+    }
+    Set-Location $RootDir
+
+    #
+    #Creating a file for snapshot
+    #
+    $sts = RunLinuxCmd -username $VMUserName -password $VMPassword -ip $Ipv4 -port $VMPort `
+    -command "touch /home/$VMUserName/PostSnapData.txt"
+
+    # Run the remote script
```
Comment (chvalean) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: fix these comments

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        Start-Sleep -S 3
+        Start-Sleep -S 3
+        $timeout -= 3
+    }
+
+    if ($timeout -eq 0) {
+        LogErr "Test case timed out waiting for VM to boot!"
+        return "FAIL"
+    }
+    # Verify VM Is ok
```
Comment (chvalean) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: please add an accurate description

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    # Check that the VM is present on the server and it is in running state.
+    #
+    $vm = Get-VM -Name $VMName -ComputerName $HvServer
+    if (-not $vm) {
+        LogErr "Cannot find VM ${vmName} on server ${hvServer}"
+        return "FAIL"
+    }
+
+
+    LogMsg "VM ${vmName} is present on server and running"
```
Comment (chvalean) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: remove this message, not relevant

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    #
+    $ExportedVM = Get-VM -Name $VMName -ComputerName $HvServer
+    $ExportedVMID = $ExportedVM.VMId
+
+    #
+    # Import back the above exported VM
+    #
+    $osInfo = Get-WmiObject Win32_OperatingSystem -ComputerName $HvServer
+    if (-not $osInfo) {
+        LogErr "Unable to collect Operating System informatioin"
```
Comment (chvalean) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: typo in information

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        return "FAIL"
+    }
+
+    [int]$BuildNum = [convert]::ToInt32($osInfo.BuildNumber, 10)
+    switch ($BuildNum) {
+        {$_ -lt 10000} { # Server 2012 R2, 20012 and 2008R2
+            $vmConfig = Get-Item "$vmPath\Virtual Machines\*.xml"
+        }
+        {$_ -ge 10000} { # TH
```
Comment (chvalean) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: correct comment to say WS2016 or newer

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+    LogMsg $vmConfig.fullname
+
+    Import-VM -Path $vmConfig -ComputerName $HvServer -Copy "${vmPath}\Virtual Hard Disks" -Verbose -Confirm:$False -GenerateNewId
+    if ($? -ne "True") {
+        LogErr "Error while importing the VM"
+        return "FAIL"
+    }
+
+    LogMsg "VM ${vmName} has imported back successfully"
```
Comment (chvalean) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: has been*

File: XML/TestCases/FunctionalTests.xml [hunk 11]
```diff
@@ -1117,4 +1117,32 @@
...
+    <test>
+        <testName>STORAGE-TAKE-REVERT-SNAPSHOT</testName>
+        <setupScript>.\TestScripts\Windows\\AddVhdxHardDisk.ps1</setupScript>
+        <testScript>SETUP-STOR-TakeRevert-Snapshot.ps1</testScript>
+        <files>.\Testscripts\Linux\utils.sh,.\Testscripts\Linux\check_traces.sh,.\Testscripts\Linux\STOR-Lis-Disk.sh</files>
+        <setupType>SingleVM</setupType>
+        <TestParameters>
+            <param>DiskType=VHD</param>
+            <param>SCSI=1,1,Dynamic</param>
+            <param>SnapshotName=ICABase</param>
```
Comment (chvalean) [inline XML/TestCases/FunctionalTests.xml:1129]: why is this required?

**Review by chvalean — CHANGES_REQUESTED** [review]

File: Testscripts/Linux/STOR-Lis-Disk.sh (hunk 1 above)
Comment (asofron) [reply Testscripts/Linux/STOR-Lis-Disk.sh]: Formated whole sh

File: Testscripts/Linux/STOR-Lis-Disk.sh (hunk 2 above)
Comment (asofron) [reply Testscripts/Linux/STOR-Lis-Disk.sh]: Made it also ignore sdb

File: Testscripts/Linux/STOR-Lis-Disk.sh (hunk 3 above)
Comment (asofron) [reply Testscripts/Linux/STOR-Lis-Disk.sh]: Update

File: Testscripts/Linux/STOR-Lis-Disk.sh (hunk 4 above)
Comment (asofron) [reply Testscripts/Linux/STOR-Lis-Disk.sh]: Done

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 (hunk 5 above)
Comment (asofron) [reply Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: Fixed

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 (hunk 6 above)
Comment (asofron) [reply Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: Added a more detailed description

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 7 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: Removed

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 8 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: Fixed

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 9 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: Updated

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 10 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: Updated

File: XML/TestCases/FunctionalTests.xml (hunk 11 above)
Comment (asofron) [reply XML/TestCases/FunctionalTests.xml:1129]: SnapshotName=ICABase is the  given name to snapshot before the revert test

File: Testscripts/Linux/STOR-Lis-Disk.sh [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    LogMsg "Checking ${blocks} blocks"
+    for ((y = 0; y < $blocks; y++)); do
+        LogMsg "Writing block $y to device $targetDevice ..."
+        dd if=$testFile of=$targetDevice bs=$blockSize count=1 seek=$y status=noxfer 2>/dev/null
+        LogMsg "Checking block $y ..."
+        testChecksum=$(dd if=$targetDevice bs=$blockSize count=1 skip=$y status=noxfer 2>/dev/null | sha1sum | cut -d " " -f 1)
+        if [ "$checksum" == "$testChecksum" ]; then
+            LogMsg "Checksum matched for block $y"
+        else
+            LogMsg "Checksum mismatch on  block $y for ${targetDevice} "
```
Comment (adriansuhov) [inline Testscripts/Linux/STOR-Lis-Disk.sh]: Pair LogErr  with SetTestStateFailed or SetTestStateAborted

File: Testscripts/Linux/STOR-Lis-Disk.sh [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+# For each drive, run fdisk -l and extract the drive
+# size in bytes.  The setup script will add Fixed
+#.vhd of size 1GB, and Dynamic .vhd of 137GB
+#
+FixedDiskSize=1073741824
+Disk4KSize=4096
+DynamicDiskSize=136365211648
+
+for driveName in /dev/sd*[^0-9]; do
+    #
```
Comment (adriansuhov) [inline Testscripts/Linux/STOR-Lis-Disk.sh]: Leave only the comment line and erase just the #

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+    #######################################################################
+    #
+    # Checking the mandatory testParams
+    #
+    $params = $TestParams.Split(";")
+    foreach ($p in $params) {
+        $fields = $p.Split("=")
+        switch ($fields[0].Trim()) {
+            "Ipv4" { $Ipv4 = $fields[1].Trim() }
+            "RootDir" { $RootDir = $fields[1].Trim() }
```
Comment (mbivolan) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: remove Ipv4 and RootDir

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+    if (-not $RootDir) {
+        "Error: Missing testParam rootDir value"
+        return "FAIL"
+    }
+
+    if (-not $Ipv4) {
+        "Error: Missing testParam ipv4 value"
+        return "FAIL"
+    }
```
Comment (mbivolan) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: Remove the checks

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    While ((Get-VM -ComputerName $HvServer -Name $newName).State -eq "On") {
+        Start-Sleep -Seconds 5
+    }
+
+    do {
+        Start-Sleep -Seconds 5
+    } until ((Get-VMIntegrationService $newName | Where-Object {$_.name -eq "Heartbeat"}).PrimaryStatusDescription -eq "OK")
+
+    LogMsg "Imported VM ${newName} has a snapshot TestExport, applied the snapshot and VM started successfully"
+
```
Comment (adriansuhov) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: Double new line

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    if (-not $snapshotname) {
+        "Error: Missing testParam snapshotname value"
```
Comment (mbivolan) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: LogErr and remove the "Error" part

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+    if ($timeout -eq 0) {
+        LogErr "Test case timed out waiting for VM to boot!"
+        return "FAIL"
+    }
+    #
+    # Verify that VM started OK after snapshot restore
+    #
+    LogMsg "Waiting for VM to boot ..."
+    $resultVMHeartbeat = Wait-VMHeartbeatOK -VMName $VMName -HvServer $HvServer `
+    -RetryCount 60 -RetryInterval 4
```
Comment (mbivolan) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: indentation

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+    LogMsg "Checking if the test file is still present..."
+    $sts = RunLinuxCmd -username $VMUserName -password $VMPassword -ip $newIpv4 -port $VMPort `
+    -ignoreLinuxExitCode -command  "stat /home/$VMUserName/PostSnapData.txt 2>/dev/null"
```
Comment (mbivolan) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: ditto

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+    if ( $sts) {
+        LogErr "File still present in VM"
+        return "FAIL"
+    }
+    return "PASS"
```
Comment (mbivolan) [inline Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: you are already returning here, all the following line will not be executed

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+    $testCaseTimeout = 600
+    #####################################################################
+    #
+    # Main script body
+    #
+    #####################################################################
+
+    # Check input arguments
+    if ($VMName -eq $null) {
+        log "VM name is null"
```
Comment (mbivolan) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: log?

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+    $params = $TestParams.Split(';')
+    foreach ($p in $params) {
+        if ($p.Trim().Length -eq 0) {
+            continue
+        }
+
+        $tokens = $p.Trim().Split('=')
+
+        if ($tokens.Length -ne 2) {
+            LogWarn "Warn: test parameter '$p' is being ignored because it appears to be malformed"
```
Comment (mbivolan) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: remove "Warn:" part

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+            continue
+        }
+
+        if ($tokens[0].Trim() -eq "RootDir") {
+            $RootDir = $tokens[1].Trim()
```
Comment (mbivolan) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: dont check for RootDir

File: Testscripts/Windows/STOR-ExportImportVM.ps1 [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+        }
+    }
+
+    # Change the working directory for the log files
+    if (-not (Test-Path $RootDir)) {
+        "Error: The directory `"${rootDir}`" does not exist"
```
Comment (mbivolan) [inline Testscripts/Windows/STOR-ExportImportVM.ps1]: LogErr

File: Testscripts/Linux/STOR-Lis-Disk.sh (hunk 13 above)
Comment (asofron) [reply Testscripts/Linux/STOR-Lis-Disk.sh]: Fixed

File: Testscripts/Linux/STOR-Lis-Disk.sh (hunk 12 above)
Comment (asofron) [reply Testscripts/Linux/STOR-Lis-Disk.sh]: Done

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 16 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: removed

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 (hunk 14 above)
Comment (asofron) [reply Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: removed

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 (hunk 17 above)
Comment (asofron) [reply Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: Done

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 (hunk 19 above)
Comment (asofron) [reply Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: Done

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 (hunk 20 above)
Comment (asofron) [reply Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: Done,removed

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 21 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: LogMsg

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 22 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: Removed

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 23 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: Done

File: Testscripts/Windows/STOR-ExportImportVM.ps1 (hunk 24 above)
Comment (asofron) [reply Testscripts/Windows/STOR-ExportImportVM.ps1]: Done

File: Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1 (hunk 15 above)
Comment (asofron) [reply Testscripts/Windows/SETUP-STOR-TakeRevert-Snapshot.ps1]: removed

File: Testscripts/Linux/STOR-Lis-Disk.sh [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+                    ls /mnt/Example
+                    rm -f /mnt/Example/data
+                    df -h
+                    umount /mnt
+                    if [ "$?" = "0" ]; then
+                        LogMsg "Drive unmounted successfully..."
+                    fi
+                    LogMsg "Disk test's completed for ${driveName}1"
+                else
+                    LogMsg "Error in creating directory /mnt/Example..."
```
Comment (chvalean) [inline Testscripts/Linux/STOR-Lis-Disk.sh]: LogErr for all these failures and below?

Comment (asofron) [reply Testscripts/Linux/STOR-Lis-Disk.sh]: Updated
